### PR TITLE
performance improve of keeper sizeInBytes calculation.

### PR DESCRIPTION
**Pull request by zhanglistar**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve keeper writing performance by optimization the size calculation logic.

Detailed description / Documentation draft:
1. add a field to track the size of Node.
2. the result is as follows:
master keeper write pressure result:
wecom-temp-a8371f90ae6514e67522caf5682a62b3

after size optimazation:
wecom-temp-23a8b987fc18b6c21fcfe64328ab9f49
ioutils of disk is 100%, so the bottle neck is disk io.

zookeeper 3.7 write pressure result:
wecom-temp-87048ffff952b2ba5a9e69e9f3eee2da
ioutils of disk is 100%, so the bottle neck is disk io.

In other machine with ssd disk, keeper is slower about 20% than zookeeper.
wecom-temp-c833436ad04aabddde2b1a9cf8773e9e

wecom-temp-54041b8f79edc983192e160aabf5783a

### CODE REVIEW

**Comment by alesapin** [comment]
@mergifyio update

**Comment by zhanglistar** [comment]
> > update
> 
> #### ❌ Pull request can't be updated with latest base branch changes

Did not find how to change this. There is no such config to set this in my page.

**Comment by alesapin** [comment]
@zhanglistar I don't know too. Just merge this branch with master manually and push.

File: src/Coordination/KeeperSnapshotManager.cpp [hunk 1]
```diff
@@ -124,6 +126,7 @@ namespace
         readBinary(node.stat.numChildren, in);
         readBinary(node.stat.pzxid, in);
         readBinary(node.seq_num, in);
+        readBinary(node.size_bytes, in);
```
Comment (tavplubix) [inline src/Coordination/KeeperSnapshotManager.cpp:129]: Should we change changelog category to Backward Incompatible Change?

Comment (zhanglistar) [reply src/Coordination/KeeperSnapshotManager.cpp:129]: > Should we change changelog category to Backward Incompatible Change?

I think there is no need to do compatible change, because the clickhouse-keeper is not release in production now, it is ongoing development state now.

Comment (alesapin) [reply src/Coordination/KeeperSnapshotManager.cpp:129]: No, we should just update snapshot version and check it on snapshot read.

Comment (zhanglistar) [reply src/Coordination/KeeperSnapshotManager.cpp]: Done.

**Comment by alesapin** [comment]
@mergifyio update

**Comment by alesapin** [comment]
@zhanglistar Sorry, we have problems with CI. can you merge with master one more time.

**Comment by zhanglistar** [comment]
> @zhanglistar Sorry, we have problems with CI. can you merge with master one more time.

Done.

**Comment by zhanglistar** [comment]
@alesapin  The failed checks cannot see the details as image show
image

File: src/Coordination/KeeperStorage.cpp [hunk 2]
```diff
@@ -133,20 +133,20 @@ static bool fixupACL(
...
-
-    total_size += data.size();
-
-    total_size += sizeof(acl_id);
-    total_size += sizeof(is_sequental);
-    total_size += sizeof(stat);
-    total_size += sizeof(seq_num);
-    return total_size;
-}
+//uint64_t KeeperStorage::Node::sizeInBytes() const
```
Comment (alesapin) [inline src/Coordination/KeeperStorage.cpp:136]: Remove this code?

File: src/Coordination/KeeperStorage.cpp [hunk 3]
```diff
@@ -524,6 +526,7 @@ struct KeeperStorageRemoveRequestProcessor final : public KeeperStorageRequestPr
                 --parent.stat.numChildren;
                 ++parent.stat.cversion;
                 parent.children.erase(child_basename);
+                parent.size_bytes -= child_basename.size();
```
Comment (alesapin) [inline src/Coordination/KeeperStorage.cpp:529]: But what about `set` request when `size` of `data` changes?

**Review by alesapin — COMMENTED** [review]
We need also update size in `set` requests because size of stored data changes. Also `setACL` request can slightly change data size.

File: src/Coordination/KeeperStorage.cpp (hunk 2 above)
Comment (zhanglistar) [reply src/Coordination/KeeperStorage.cpp]: done

File: src/Coordination/KeeperStorage.cpp (hunk 3 above)
Comment (zhanglistar) [reply src/Coordination/KeeperStorage.cpp:514]: done

Comment (zhanglistar) [reply src/Coordination/KeeperStorage.cpp:514]: There is no need to update size in set acl processor because of acl is fixed width.

**Comment by zhanglistar** [comment]
@alesapin  I have checked  the checks result has non relationship with keeper.

**Comment by zhanglistar** [comment]
> @alesapin I have checked the checks that result has non relationship with keeper. Can merge?

@alexey-milovidov @alesapin can merge?

File: src/Coordination/KeeperStorage.cpp [hunk 4]
```diff
@@ -621,11 +610,11 @@ struct KeeperStorageSetRequestProcessor final : public KeeperStorageRequestProce
 
             auto itr = container.updateValue(request.path, [zxid, request] (KeeperStorage::Node & value)
             {
-                value.data = request.data;
                 value.stat.version++;
                 value.stat.mzxid = zxid;
                 value.stat.mtime = std::chrono::system_clock::now().time_since_epoch() / std::chrono::milliseconds(1);
                 value.stat.dataLength = request.data.length();
+                value.size_bytes += request.data.size() - value.data.size();
```
Comment (alesapin) [inline src/Coordination/KeeperStorage.cpp:617]: What if this diff is negative? Probably we will get an overflow subtracting two unsigned integers https://www.cplusplus.com/reference/string/string/size/?

Comment (zhanglistar) [reply src/Coordination/KeeperStorage.cpp]: fix.

Comment (zhanglistar) [reply src/Coordination/KeeperStorage.cpp]: > 

That's really considerate.

File: src/Parsers/tests/gtest_Parser.cpp [hunk 5]
```diff
@@ -157,7 +157,7 @@ SOURCE(CLICKHOUSE(HOST 'localhost' PORT tcpPort() TABLE 'source_table'))
 LAYOUT(FLAT())
 LIFETIME(MIN 0 MAX 1000)
 COMMENT 'Test dictionary with comment';
-)sql",
+sql",
```
Comment (cmsxbc) [inline src/Parsers/tests/gtest_Parser.cpp:160]: isn't it a unnecessary change?

Comment (zhanglistar) [reply src/Parsers/tests/gtest_Parser.cpp:160]: > isn't it a unnecessary change?

it is a unit test fail, so I fix it by the way.

Comment (cmsxbc) [reply src/Parsers/tests/gtest_Parser.cpp:160]: > > isn't it a unnecessary change?
> 
> it is a unit test fail, so I fix it by the way.


No, it does not matter. What you changed is C++ Raw String, the `)` you  removed is part for spliting content and delimiters.  Althought I don't know why there is no compiler warning/error when you removed it.

A helpful document [string_literal](https://en.cppreference.com/w/cpp/language/string_literal)

Comment (cmsxbc) [reply src/Parsers/tests/gtest_Parser.cpp:160]: And the unit test failed really strange. But it didn't occur in recently master. So, it's not necessary to change this test. maybe you can try to merge master again.

Comment (zhanglistar) [reply src/Parsers/tests/gtest_Parser.cpp:160]: > And the unit test failed really strange. But it dose not occur in recently master. So, it's not necessary to change this test. maybe you can try to merge master again.

That's wired, I change the code and the test passes. Maybe need deep the reason.

Comment (cmsxbc) [reply src/Parsers/tests/gtest_Parser.cpp:201]: I think this is correctly fix for this ut.
https://github.com/ClickHouse/ClickHouse/commit/bb521cd363f56c040cdbba523e48bc8496956c62#diff-29a9dc03cc52b629c442735c065033c77fa61e1ae7504191adad7f5ef47cdce2R188

Comment (zhanglistar) [reply src/Parsers/tests/gtest_Parser.cpp:201]: I rollback this code.

**Review by alesapin — APPROVED** [review]

**Comment by alesapin** [comment]
Unfortunately all failures unrelated to changes.

**Comment by cmsxbc** [comment]
@alesapin I think you have merge a mistake, I just create another pr to fix it. #32912
